build.sbt

```
@@ -8,17 +8,95 @@ import org.scalajs.linker.interface.ModuleSplitStyle
-ThisBuild / dockerUpdateLatest := true
-
 ThisBuild / resolvers := List(Resolver.mavenLocal)
 
+val pushCond          = "github.event_name == 'push'"
+val prCond            = "github.event_name == 'pull_request'"
+val mainCond          = "github.ref == 'refs/heads/main'"
+val notMainCond       = "github.ref != 'refs/heads/main'"
+val geminiRepoCond    = "startsWith(github.repository, 'gemini')"
+val notDependabotCond = "github.actor != 'dependabot[bot]'"
+val isMergedCond      = "github.event.pull_request.merged == true"
+def allConds(conds: String*) = conds.mkString("(", " && ", ")")
+def anyConds(conds: String*) = conds.mkString("(", " || ", ")")
+
+val faNpmAuthToken = "FONTAWESOME_NPM_AUTH_TOKEN" -> "${{ secrets.FONTAWESOME_NPM_AUTH_TOKEN }}"
+val herokuToken    = "HEROKU_API_KEY"             -> "${{ secrets.HEROKU_API_KEY }}"
+
 ThisBuild / githubWorkflowSbtCommand := "sbt -v -J-Xmx6g"
+ThisBuild / githubWorkflowEnv += faNpmAuthToken
+ThisBuild / githubWorkflowEnv += herokuToken
+
+lazy val setupNodeNpmInstall =
+  List(
+    WorkflowStep.Use(
+      UseRef.Public("actions", "setup-node", "v4"),
+      name = Some("Setup Node.js"),
+      params = Map(
+        "node-version"          -> "20",
+        "cache"                 -> "npm",
+        "cache-dependency-path" -> "modules/web/client/package-lock.json"
+      )
+    ),
+    WorkflowStep.Use(
+      UseRef.Public("actions", "cache", "v3"),
+      name = Some("Cache node_modules"),
+      id = Some("cache-node_modules"),
+      params = {
+        val prefix = "node_modules"
+        val key    = s"$prefix-$${{ hashFiles('modules/web/client/package-lock.json') }}"
+        Map("path" -> "node_modules", "key" -> key, "restore-keys" -> prefix)
+      }
+    ),
+    WorkflowStep.Run(
+      List("cd modules/web/client", "npm clean-install --verbose"),
+      name = Some("npm clean-install"),
+      cond = Some("steps.cache-node_modules.outputs.cache-hit != 'true'")
+    )
+  )
+
+lazy val herokuLogin =
+  WorkflowStep.Run(
+    List("heroku container:login"),
+    name = Some("Login to Heroku Container")
+  )
+
+lazy val sbtDeploy =
+  WorkflowStep.Sbt(
+    List("deploy_observe_server_staging/docker:publish"),
+    name = Some("Build and deploy Docker image to Heroku")
+  )
+
+lazy val herokuRelease =
+  WorkflowStep.Run(
+    List("heroku container:release web -a ${{ secrets.HEROKU_APP_NAME }}"),
+    name = Some("Release app in Heroku")
+  )
+
+ThisBuild / githubWorkflowAddedJobs +=
+  WorkflowJob(
+    "deploy",
+    "Deploy staging app to Heroku",
+    WorkflowStep.Checkout ::
+      WorkflowStep.SetupJava(githubWorkflowJavaVersions.value.toList.take(1)) :::
+      setupNodeNpmInstall :::
+      herokuLogin ::
+      sbtDeploy ::
+      herokuRelease ::
+      Nil,
+    scalas = List(scalaVersion.value),
+    javas = githubWorkflowJavaVersions.value.toList.take(1),
+    cond = Some(allConds(anyConds(mainCond, prCond), geminiRepoCond, isMergedCond))
```

I think this is all we need. Since we only want to publish commits to the main branch, is that right?

```suggestion
    cond = Some(allConds(mainCond, geminiRepoCond))
```

---

Is `mainCond` true when it runs on the main branch, or when the PR wants to merge to the main branch too?

---

Let's try it.

---

`mainCond` is true only when the commit is directly pushed to the `main` branch (i.e. in a merge commit). When a PR wants to merge into main, then all its commits must exist on a non-`main` branch, and therefore fail this condition.